rust-sdk/whirlpool/src/token.rs

```diff
@@ -333,4 +342,233 @@ mod tests {
         assert_eq!(mint_c, mint1);
         assert_eq!(mint_d, mint2);
     }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_token_2022_extensions() {
+        let ctx = RpcContext::new().await;
+
+        // Create Token-2022 mint with transfer fee
+        let mint_te = setup_mint_te_fee(&ctx).await.unwrap();
+        let mint_account = ctx.rpc.get_account(&mint_te).await.unwrap();
+
+        // Test transfer fee at epoch 0
+        let older = get_current_transfer_fee(Some(&mint_account), 0).unwrap();
+        assert_eq!(older.fee_bps, 100); // 1%
+        assert_eq!(older.max_fee, 1_000_000_000); // 1 token
+
+        // Test transfer fee at epoch 2
+        let newer = get_current_transfer_fee(Some(&mint_account), 2).unwrap();
+        assert_eq!(newer.fee_bps, 150); // 1.5%
+        assert_eq!(newer.max_fee, 1_000_000_000); // 1 token
+
+        // Test with no fee
+        let no_fee_result = get_current_transfer_fee(None, 0);
+        assert!(no_fee_result.is_none());
+    }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_token_2022_account() {
+        let ctx = RpcContext::new().await;
+
+        // Create basic Token-2022 mint (without transfer fee)
+        let mint = setup_mint_te(&ctx, &[]).await.unwrap();
+        let mint_account = ctx.rpc.get_account(&mint).await.unwrap();
+
+        // Verify account data
+        assert_eq!(mint_account.data.len(), 82);
+        assert_eq!(mint_account.owner, TOKEN_2022_PROGRAM_ID);
+    }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_token_2022_with_transfer_fee() {
+        let ctx = RpcContext::new().await;
+
+        // Create Token-2022 mint with transfer fee
+        let mint = setup_mint_te_fee(&ctx).await.unwrap();
+
+        // Verify account data
+        let account = ctx.rpc.get_account(&mint).await.unwrap();
+        assert!(account.data.len() > 82); // Size is larger due to extension
+        assert_eq!(account.owner, TOKEN_2022_PROGRAM_ID);
+    }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_no_tokens() {
+        let ctx = RpcContext::new().await;
+        let result = prepare_token_accounts_instructions(&ctx.rpc, ctx.signer.pubkey(), vec![])
+            .await
+            .unwrap();
+
+        assert_eq!(result.create_instructions.len(), 0);
+        assert_eq!(result.cleanup_instructions.len(), 0);
+        assert_eq!(result.token_account_addresses.len(), 0);
+    }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_native_mint_wrapping_none() {
+        let ctx = RpcContext::new().await;
+        crate::set_native_mint_wrapping_strategy(NativeMintWrappingStrategy::None).unwrap();
+
+        let result = prepare_token_accounts_instructions(
+            &ctx.rpc,
+            ctx.signer.pubkey(),
+            vec![TokenAccountStrategy::WithoutBalance(native_mint::ID)],
+        )
+        .await
+        .unwrap();
+
+        assert_eq!(result.create_instructions.len(), 1); // Create ATA
```

Comment:
This right here is only asserting the length of the instructions which only checks if there are instructions but does not check if said instructions are correct.

What I would do:
* Take the BanksClient state before `does the token account for NATIVE_MINT exist?`
* Simulate the create instructions using BanksClient (`ctx.send_transaction_with_signers` or `ctx.send_transaction`).
* Take the BanksClient state after `does the token account for NATIVE_MINT exist now?`
* Simulate the cleanup instructions using BanksClient
* Take the BanksClient state after `does the token account for NATIVE_MINT exist now?`

This way:
* We know that the instructions are valid cause they can be simulated by BanksClient
* Know that the instructions do what we want them to do based on the outcome of the instruction: `initialize token account(s) with or without balance checks`.

Reply:
Ok updated!